src/analysis/souffle/sensitivity_analysis.dl

```diff
@@ -0,0 +1,132 @@
+  isSqlValue(value), sqlValueHasSqlInputList(value, input_list),
+  sqlOperatorInputListLength(input_list, len).
+
+//indexing the input per value 
+.decl sqlValueHasSqlInputAtIndex(value: AccessPath, sql_input: AccessPath, index:number)
+
+sqlValueHasSqlInputAtIndex(value, sql_input, sql_input_list_len - suffix_list_len) :-
+       sqlValueHasSqlInputListSuffix(value, [sql_input, tail]), 
+       sqlOperatorInputListLength([sql_input, tail], suffix_list_len), 
+       sqlValueSqlOperatorInputListLength(value, sql_input_list_len). 
+
+//keeping track of all initial inputs as they all have sennsitivity score 1
+.decl isInitialInput (ingress_point: AccessPath)
+
+//helper functions to flatten input lists
+.decl intermediateFlatten(list:OperandList, sublist:OperandList, list_element:AccessPath)
+
+intermediateFlatten(list, sublist, list_element) :- isSqlOperator([_, _, _, list, _]), list = [list_element, sublist].
+intermediateFlatten(list, sublist, list_element) :- intermediateFlatten(list, [list_element, sublist], _).
+
+.decl flatten(list:OperandList, list_element:AccessPath)
+flatten(list, list_element) :- intermediateFlatten(list, _, list_element).
+
+//flattening each individual input per value
+.decl isSqlInput(value: AccessPath, sql_input: AccessPath)
+
+isSqlInput(value, sql_input) :- isSqlOperator([_, _, value, input_list, _]), 
+                              flatten(input_list, sql_input). 
+
+//keeping track of sensitivity score per each input at index 
+//used to calculate the maximum input sensitivity 
+.decl sqlValueHasSensitivityScoreAtIndex(value: AccessPath, sensitivity_score: number, index: number) 
+
+//base cases (index 0)
+//for initial inputs, sensitivity is 1
+sqlValueHasSensitivityScoreAtIndex(value, 1, 0) :- sqlValueHasSqlInputAtIndex(value, sql_input, 0),
+      isInitialInput(sql_input).
+
+//for all others, get the final sensitivity score
+sqlValueHasSensitivityScoreAtIndex(value, sensitivity_score, 0) :- sqlValueHasSqlInputAtIndex(value, sql_input, 0),
+       sqlValueSensitivity(sql_input, sensitivity_score).      
+
+//inductive cases 
+//keep track at the score at each index, only update if the new input score is higher
+sqlValueHasSensitivityScoreAtIndex(value, sensitivity_score, index) :- 
+      isSqlValue(value),
+      sqlValueHasSqlInputAtIndex(value, sql_input, index), 
+      isSqlInput(value, sql_input), 
+      sqlValueSensitivity(sql_input, input_sensitivity_score), 
+      sqlValueHasSensitivityScoreAtIndex(value, sensitivity_score, index-1),
+      sensitivity_score >= input_sensitivity_score. 
+
+
+sqlValueHasSensitivityScoreAtIndex(value, sensitivity_score, index) :- 
+      isSqlValue(value),
+      sqlValueHasSqlInputAtIndex(value, sql_input, index), 
+      isSqlInput(value, sql_input), 
+      sqlValueSensitivity(sql_input, sensitivity_score), 
+      sqlValueHasSensitivityScoreAtIndex(value, prev_sensitivity_score, index-1), 
+      sensitivity_score >= prev_sensitivity_score. 
```

Comment:
I'm pretty confused by the relationship between these two relations. Are they trying to do the same thing, or something slightly different?

Reply:
They're supposed to do slightly different things (see last line). The first one makes sure the sensitivity score is greater or equal then the sensitivity score of the input at  *index*; the second one makes sure the sensitivity score is greater or equal than the previous sensitivity score; combined, they should update the current score only if the sensitivity score of input at *index* is greater than the previous sensitivity score (*index-1*); otherwise, just copy it. My souffle knowledge is still not very well formed so please let me know if you think I should change anything or if the code is plain wrong. Alternatively, would adding a comment explaining what I just did help? 

Sidenote: Gogul mentioned I might be able to combine this in one line using an *or* operator. Would that help?